cartographer_grpc/framework/rpc.cc

```diff
@@ -130,86 +131,37 @@ void Rpc::RequestStreamingReadIfNeeded() {
 }
 
 void Rpc::Write(std::unique_ptr<::google::protobuf::Message> message) {
-  switch (rpc_handler_info_.rpc_type) {
-    case ::grpc::internal::RpcMethod::BIDI_STREAMING:
-      // For BIDI_STREAMING enqueue the message into the send queue and
-      // start write operations if none are currently in flight.
-      send_queue_.emplace(SendItem{std::move(message), ::grpc::Status::OK});
-      PerformWriteIfNeeded();
-      break;
-    case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
-      SendFinish(std::move(message), ::grpc::Status::OK);
-      break;
-    case ::grpc::internal::RpcMethod::NORMAL_RPC:
-      SendFinish(std::move(message), ::grpc::Status::OK);
-      break;
-    default:
-      LOG(FATAL) << "RPC type not implemented.";
-  }
-}
-
-void Rpc::SendFinish(std::unique_ptr<::google::protobuf::Message> message,
-                     ::grpc::Status status) {
-  SetRpcEventState(Event::FINISH, true);
-  switch (rpc_handler_info_.rpc_type) {
-    case ::grpc::internal::RpcMethod::BIDI_STREAMING:
-      CHECK(!message);
-      server_async_reader_writer_->Finish(
-          status, new RpcEvent{Event::FINISH, weak_ptr_factory_(this)});
-      break;
-    case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
-      response_ = std::move(message);
-      SendUnaryFinish(server_async_reader_.get(), status, response_.get(),
-                      new RpcEvent{Event::FINISH, weak_ptr_factory_(this)});
-      break;
-    case ::grpc::internal::RpcMethod::NORMAL_RPC:
-      response_ = std::move(message);
-      SendUnaryFinish(server_async_response_writer_.get(), status,
-                      response_.get(),
-                      new RpcEvent{Event::FINISH, weak_ptr_factory_(this)});
-      break;
-    default:
-      LOG(FATAL) << "RPC type not implemented.";
-  }
+  EnqueueMessage(SendItem{std::move(message), ::grpc::Status::OK});
+  event_queue_->Push(
+      new RpcEvent{Event::WRITE_NEEDED, weak_ptr_factory_(this), true});
 }
 
 void Rpc::Finish(::grpc::Status status) {
-  switch (rpc_handler_info_.rpc_type) {
-    case ::grpc::internal::RpcMethod::BIDI_STREAMING:
-      send_queue_.emplace(SendItem{nullptr /* msg */, status});
-      PerformWriteIfNeeded();
-      break;
-    case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
-      SendFinish(nullptr /* message */, status);
-      break;
-    case ::grpc::internal::RpcMethod::NORMAL_RPC:
-      SendFinish(nullptr /* message */, status);
-      break;
-    default:
-      LOG(FATAL) << "RPC type not implemented.";
-  }
+  EnqueueMessage(SendItem{nullptr /* message */, status});
+  event_queue_->Push(
+      new RpcEvent{Event::WRITE_NEEDED, weak_ptr_factory_(this), true});
 }
 
 void Rpc::PerformWriteIfNeeded() {
```

Comment:
naming: Perhaps HandleSendQueue is more telling?

Reply:
Done.